### PR TITLE
[3/?] Embed art if the file wasn't converted but only copied, and ensure that updated artwork will be embedded anew.

### PR DESCRIPTION
**Pull request by wisp3rwind**
Third in a series of PRs to backport changes from  [wisp3rwind/beets-alternatives](https://github.com/wisp3rwind/beets-alternatives) to this repo

Previously, embedded artwork was not guaranteed to be up-to-date with the cover file tracked by beets. The patch includes a little rework of the `item_action` handling, I hope this seems a like a reasonable idea to you, too.

### CODE REVIEW

File: beetsplug/alternatives.py [hunk 1]
```diff
@@ -97,6 +97,7 @@ class External(object):
     WRITE = 3
     MOVE = 4
     NOOP = 5
```
Comment (geigerzaehler) [inline beetsplug/alternatives.py]: We can remove this now. It is unused

File: beetsplug/alternatives.py [hunk 2]
```diff
@@ -125,19 +126,24 @@ def parse_config(self, config):
 
     def matched_item_action(self, item):
         path = self.get_path(item)
+        actions = []
         if path and os.path.isfile(path):
             dest = self.destination(item)
             if path != dest:
-                return (item, self.MOVE)
+                actions.extend([self.MOVE, self.WRITE])
```
Comment (geigerzaehler) [inline beetsplug/alternatives.py]: I guess the `self.WRITE` action is here so that we cover the case when a file has been moved _and_ updated, correct?

Couldn’t we achieve the same thing with this code?
~~~python
if path != dest:
    actions.apped(self.MOVE)
if (os.path.getmtime(syspath(dest))
        < os.path.getmtime(syspath(item.path))):
    actions.append(self.WRITE)
~~~

Comment (wisp3rwind) [reply beetsplug/alternatives.py]: I guess so. However, the write operation in that situation is not new actually, I only moved it from [the `update()` function](https://github.com/geigerzaehler/beets-alternatives/pull/17/files/94742e403a049d80571e9746f26c6bc0c5c41523#diff-ebb574ba2f6ca42686b2991565c7dbe2L183) to `matched_item_action()`. I guess that was to disentangle the `MOVE` and `WRITE` operation (it is quite some time ago that I initially made this commit...) and certainly unrelated to the title of this PR.

To make your code suggestion work, it should check the file at `path` instead of the one at `dest`, though:
```python
if path != dest:
    actions.append(self.MOVE)
if (os.path.getmtime(syspath(path))
        < os.path.getmtime(syspath(item.path))):
    actions.append(self.WRITE)
```

That is probably more concise than the current code. There might be some (not so clear) implication on performance, although I think the limiting factor is likely to beets database performance anyway:
* if the `MOVE` is caused by a change to the path formats (i.e. the templates), your version could prevent unnecessary `WRITE`s since it checks mtime,
* if the `MOVE` is triggered by changed metadata (title, ... , whatever is used by the template), the `WRITE` is required and the original code saved on filesystem accesses by not calling `getmtime`.

I think I'm going to use the code you suggested.

File: beetsplug/alternatives.py (hunk 1 above)
Comment (wisp3rwind) [reply beetsplug/alternatives.py]: :+1:

**Comment by wisp3rwind** [comment]
Ok, I'll rebase this on master and have a look at the testing framework

File: beetsplug/alternatives.py (hunk 2 above)
Comment (geigerzaehler) [reply beetsplug/alternatives.py]: You’re right about `path` instead of `dest` and spot on with regard to the performance trade-off.

**Comment by wisp3rwind** [comment]
Just force-pushed an updated patch that includes tests and hopefully is also better structured into several smaller commits. What do you think? There are some comments left in the code about issues I encountered: In particular, actually always being in sync with the source artwork is probably not worth the filesystem load it would require.

File: test/cli_test.py [hunk 3]
```diff
@@ -219,6 +224,68 @@ def test_unkown_collection(self):
...
+        def rm_art():
+            try:
+                os.remove(fname)
+            except OSError:
+                # If the test succeeds, the file will have been removed already
+                pass
+
+        fh, fname = tempfile.mkstemp()
+        os.close(fh)
+        self.addCleanup(rm_art)
```
Comment (geigerzaehler) [inline test/cli_test.py]: You can use `mkdtemp` from `TestHelper`. It will take care of cleaning up the directory on teardown.
~~~python
image_dir = self.mkdtemp()
image_path = os.path.join(image_dir, 'image')
shutil.copy(self.IMAGE_FIXTURE1, image_path)
~~~

**Comment by wisp3rwind** [comment]
> I’d be ok with paying a performance penalty for this feature. We already check a lot of mtimes and we can follow-up with performance improvements later.

I'm not quite sure if I understand you correctly here: You're saying that checking a few additional mtimes is ok, or that it would be fine to also do the `MediaFile` check for actual embedded art? I didn't benchmark the latter (or anything for that matter), but I would assume that it's a significant slowdown due to accessing file contents instead of filesystem metadata only.
One approach could be to drop the artwork removal codepaths and the comment about caching the result of `if (album.artpath and os.path.isfile(album.artpath) [...])` for now, but open another PR/issue with them to track the issue?

> Once you address the comments and fix the test this should be ok to merge.

That might warrant some thorough investigation... I've seen this test failure before, but then it disappeared. `tox` succeed locally before I pushed the last changes. The only reason that I can think of right now is some non-deterministic padding being applied to the image data.

**Comment by geigerzaehler** [comment]
> You're saying that checking a few additional mtimes is ok, or that it would be fine to also do the `MediaFile` check for actual embedded art?

The first one. I think your right that checking the `MediaFile` would be costly. That means we would ignore the artwork remove edge case for now and track in in an issue.

I’m going to have a look at the test failures later.

**Comment by wisp3rwind** [comment]
So I did change everything you commented on in your review. I couldn't reproduce the test failure, though (running tox 100 times in a loop does not change that either):

- Going through Travis' logs again, my initial suggestions of some unexpected mutagen interactions or not properly writing out the file to disk don't seem to be the problem: `beets-alternatives` does not even schedule another `SYNC_ART` after changing the artwork. If it did, the `~<filename>` line would be printed a second time.
- I couldn't find any logic bug, however. So if this is indeed related to the Travis environment, maybe the mtime handling by the OS could be the culprit? Although I don't really believe this myself, low granularity might be the problem (which I would not really expect on a Linux filesystem. I've seen such issues on FAT disks, though).

Anyway, I'm going to push the new patch. If Travis continues to fail, I'll push a version that logs the mtimes and which decisions it takes in `matched_item_action`.

**Comment by wisp3rwind** [comment]
Temporarily pushed a brunch with some more debugging instrumentation since Travis kept failing. Some local testing suggests that mtime granularity could indeed be a problem given that the tests run rather fast.
